README.md
# BoringAuth

[![Build Status](https://api.travis-ci.org/ThinkAlexandria/boringauth.svg?branch=master)](https://travis-ci.org/breard-r/boringauth)
[![LibreAuth on crates.io](https://img.shields.io/crates/v/boringauth.svg)](https://crates.io/crates/boringauth)

BoringAuth is a collection of tools for user authentication. BoringAuth is a fork
of [LibreAuth](https://github.com/breard-r/boringauth) that chooses to use the
actively developed **ring** crypto crate over the dead **rust-crypto** crate for
its crypto primitives.


## Features

- Password / passphrase authentication
  - [x] no character-set limitation
  - [x] reasonable lenth limit ([security vs. DOS](http://arstechnica.com/security/2013/09/long-passwords-are-good-but-too-much-length-can-be-bad-for-security/))
  - [x] strong, evolutive and retro-compatible password derivation functions
  - [x] crypt() compatibility
- HOTP - HMAC-based One-time Password Algorithm ([OATH](http://www.openauthentication.org/) - [RFC 4226](https://tools.ietf.org/html/rfc4226))
  - [x] the key can be passed as bytes, an ASCII string, an hexadicimal string or a base32 string
  - [x] customizable counter
  - [x] customizable hash function (sha1, sha256, sha512)
  - [x] customizable output length
  - [x] customizable output alphabet
- TOTP - Time-based One-time Password Algorithm ([OATH](http://www.openauthentication.org/) - [RFC 6238](https://tools.ietf.org/html/rfc6238))
  - [x] the key can be passed as bytes, an ASCII string, an hexadicimal string or a base32 string
  - [x] customizable timestamp
  - [x] customizable period
  - [x] customizable initial time (T0)
  - [x] customizable hash function (sha1, sha256, sha512)
  - [x] customizable output length
  - [x] customizable output alphabet
  - [x] customizable positive and negative period tolerance
- YubiKey OTP ([Yubico](https://developers.yubico.com/OTP/))
  - [ ] virtual device API
  - [ ] client API
  - [ ] server API
- U2F - Universal 2nd Factor ([FIDO Alliance](https://fidoalliance.org/specifications/download/))
  - [ ] virtual device API
  - [ ] client API
  - [ ] server API


## Using within a Rust project

You can find BoringAuth on [crates.io](https://crates.io/crates/boringauth) and include it in your `Cargo.toml`:

```toml
boringauth = "*"
```


## Using outside Rust

In order to build BoringAuth, you will need both the [rust compiler](https://github.com/rust-lang/rust) and [cargo](https://github.com/rust-lang/cargo).

```ShellSession
$ git clone https://github.com/ThinkAlexandria/boringauth.git
$ cd boringauth
$ make
$ make install prefix=/usr
```


## Quick examples


### Rust


```rust
extern crate boringauth;
use boringauth::oath::TOTPBuilder;

let key = "GEZDGNBVGY3TQOJQGEZDGNBVGY3TQOJQ".to_string();
let code = TOTPBuilder::new()
    .base32_key(&key)
    .finalize()
    .unwrap()
    .generate();
assert_eq!(code.len(), 6);
```

### C

```C
#include <stdio.h>
#include <boringauth.h>

int main(void) {
  struct boringauth_totp_cfg cfg;
  char   code[7], key[] = "12345678901234567890";

  if (boringauth_totp_init(&cfg) != LIBREAUTH_OTP_SUCCESS) {
    return 1;
  }
  cfg.key = key;
  cfg.key_len = sizeof(key);
  if (boringauth_totp_generate(&cfg, code) != LIBREAUTH_OTP_SUCCESS) {
    return 2;
  }

  printf("%s\n", code);

  return 0;
}
```

```ShellSession
$ cc -o totp totp.c -lboringauth
$ ./totp
848085
```

### Python

```Python
from ctypes.util import find_library
from struct import Struct
from ctypes import *

class TOTPcfg(Structure):
    _fields_ = [
        ('key', c_char_p),
        ('key_len', c_size_t),
        ('timestamp', c_longlong),
        ('period', c_uint),
        ('initial_time', c_ulonglong),
        ('output_len', c_size_t),
        ('output_base', c_char_p),
        ('output_base_len', c_size_t),
        ('hash_function', c_int),
    ]

def get_totp():
    key = b'12345678901234567890'
    lib_path = find_library('boringauth') or 'target/release/libboringauth.so'
    lib = cdll.LoadLibrary(lib_path)
    cfg = TOTPcfg()
    if lib.boringauth_totp_init(byref(cfg)) != 0:
        return
    cfg.key_len = len(key)
    cfg.key = c_char_p(key)
    code = create_string_buffer(b'\000' * cfg.output_len)
    if lib.boringauth_totp_generate(byref(cfg), code) != 0:
        return
    return str(code.value, encoding="utf-8")

if __name__ == '__main__':
    code = get_totp()
    print('{}'.format(code))
```


CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file.
This project adheres to [Semantic Versioning](http://semver.org/).


## [Unreleased][unreleased]

### Added
### Changed

## [0.6.0] - 2017-06-04

### Added
- Password authentication.
- Generic parsing module for PHC modular crypt format.

### Changed
- The project has been forked from LibreAuth and renamed to BoringAuth.
- The **ring** cryptographic primitives crate has replaced the **rust-crypto** crate.
- The modifications to LibreAuth together with the original work are released combined under the GPL v2.0 license.
- The oath module does not not expose sub-modules anymore.


## [0.5.2] - 2016-06-26

### Changed
- The project has been renamed LibreAuth (previous name: R2FA);
- the license has been changed to the CeCILL Free Software License Agreement v2.1;
- the otp module has been renamed oath.

### Removed
- The otp::c (oath::c) module is not public anymore.


## [0.5.1] - 2015-08-07
This release fixed a buffer overflow in the C-bindings tests. Because this issue did not influence the Rust part of the code in any way, the crate has not be updated.


## [0.5.0] - 2015-08-06

### Added
- C-bindings for OTP validation.

### Changed
- Errors are now represented by an enumeration instead of strings;
- the C unit test suite has been rewritten.

### Fixed
- An integer overflow has been fixed in code length calculation.


## [0.4.2] - 2015-08-04
This release is a cleanup release. No public interface has been changed.


## [0.4.1] - 2015-08-03

### Added
- C-bindings are now part of this library and therefore no longer distributed in a separated project.


## [0.4.0] - 2015-08-01

Last version without a changelog.


tests/test_hotp.c
/*
 * Copyright Rodolphe Breard (2015)
 * Author: Rodolphe Breard (2015)
 *
 * This software is a computer program whose purpose is to [describe
 * functionalities and technical features of your software].
 *
 * This software is governed by the CeCILL  license under French law and
 * abiding by the rules of distribution of free software.  You can  use,
 * modify and/ or redistribute the software under the terms of the CeCILL
 * license as circulated by CEA, CNRS and INRIA at the following URL
 * "http://www.cecill.info".
 *
 * As a counterpart to the access to the source code and  rights to copy,
 * modify and redistribute granted by the license, users are provided only
 * with a limited warranty  and the software's author,  the holder of the
 * economic rights,  and the successive licensors  have only  limited
 * liability.
 *
 * In this respect, the user's attention is drawn to the risks associated
 * with loading,  using,  modifying and/or developing or reproducing the
 * software by the user in light of its specific status of free software,
 * that may mean  that it is complicated to manipulate,  and  that  also
 * therefore means  that it is reserved for developers  and  experienced
 * professionals having in-depth computer knowledge. Users are therefore
 * encouraged to load and test the software's suitability as regards their
 * requirements in conditions enabling the security of their systems and/or
 * data to be ensured and,  more generally, to use and operate it in the
 * same conditions as regards security.
 *
 * The fact that you are presently reading this means that you have had
 * knowledge of the CeCILL license and that you accept its terms.
 */


#include <strings.h>
#include <string.h>
#include <assert.h>
#include <boringauth.h>
#include "boringauth_tests.h"

#define DEFAULT_BUFF_LEN  6


static uint32_t test_basic_hotp(void) {
    test_name("hotp: test_basic_hotp");

    struct boringauth_hotp_cfg cfg;
    const char key[] = "12345678901234567890";
    char code[DEFAULT_BUFF_LEN + 1];

    uint32_t ret = boringauth_hotp_init(&cfg);
    assert(ret == LIBREAUTH_OATH_SUCCESS);
    assert(cfg.key == NULL);
    assert(cfg.key_len == 0);
    assert(cfg.counter == 0);
    assert(cfg.output_len == DEFAULT_BUFF_LEN);
    assert(cfg.output_base == NULL);
    assert(cfg.output_base_len == 0);
    assert(cfg.hash_function == LIBREAUTH_OATH_SHA_1);

    cfg.key = key;
    cfg.key_len = sizeof(key);

    ret = boringauth_hotp_generate(&cfg, code);
    assert(ret == LIBREAUTH_OATH_SUCCESS);
    assert(strlen(code) == DEFAULT_BUFF_LEN);
    assert(strncmp(code, "755224", DEFAULT_BUFF_LEN + 1) == 0);

    assert(boringauth_hotp_is_valid(&cfg, "755224"));
    assert(!boringauth_hotp_is_valid(NULL, "755224"));
    assert(!boringauth_hotp_is_valid(&cfg, "755225"));
    assert(!boringauth_hotp_is_valid(&cfg, "4755224"));
    assert(!boringauth_hotp_is_valid(&cfg, "!@#$%^"));
    assert(!boringauth_hotp_is_valid(&cfg, ""));
    assert(!boringauth_hotp_is_valid(&cfg, NULL));

    return 1;
}

static uint32_t test_init_null_ptr(void) {
    test_name("hotp: test_init_null_ptr");

    uint32_t ret = boringauth_hotp_init(NULL);
    assert(ret == LIBREAUTH_OATH_CFG_NULL_PTR);

    return 1;
}

static uint32_t test_generate_null_ptr(void) {
    test_name("hotp: test_generate_null_ptr");

    struct boringauth_hotp_cfg cfg;
    const char key[] = "12345678901234567890";
    char code[] = "qwerty";
    uint32_t ret;

    boringauth_hotp_init(&cfg);

    ret = boringauth_hotp_generate(NULL, code);
    assert(ret == LIBREAUTH_OATH_CFG_NULL_PTR);
    assert(strcmp(code, "qwerty") == 0);

    ret = boringauth_hotp_generate(&cfg, code);
    assert(ret == LIBREAUTH_OATH_KEY_NULL_PTR);

    cfg.key = key;

    ret = boringauth_hotp_generate(&cfg, code);
    assert(ret == LIBREAUTH_OATH_INVALID_KEY_LEN);

    cfg.key_len = sizeof(key);

    ret = boringauth_hotp_generate(&cfg, NULL);
    assert(ret == LIBREAUTH_OATH_CODE_NULL_PTR);

    ret = boringauth_hotp_generate(&cfg, code);
    assert(ret == LIBREAUTH_OATH_SUCCESS);

    return 1;
}

static uint32_t test_invalid_base(void) {
    test_name("hotp: test_invalid_base");

    struct boringauth_hotp_cfg cfg;
    const char key[] = "12345678901234567890", base[] = "0123456789ABCDEF";
    char code[DEFAULT_BUFF_LEN + 1];

    boringauth_hotp_init(&cfg);

    cfg.key = key;
    cfg.key_len = sizeof(key);
    cfg.output_base = base;

    uint32_t ret = boringauth_hotp_generate(&cfg, code);
    assert(ret == LIBREAUTH_OATH_INVALID_BASE_LEN);
    cfg.output_base_len = 1;
    ret = boringauth_hotp_generate(&cfg, code);
    assert(ret == LIBREAUTH_OATH_INVALID_BASE_LEN);

    cfg.output_base_len = sizeof(base);

    ret = boringauth_hotp_generate(&cfg, code);
    assert(ret == LIBREAUTH_OATH_SUCCESS);

    return 1;
}

static uint32_t test_invalid_code(void) {
    test_name("hotp: test_invalid_code");

    struct boringauth_hotp_cfg cfg;
    const char key[] = "12345678901234567890",
          base10[] = "0123456789",
          base32[] = "ABCDEFGHIJKLMNOPQRSTUVWXYZ234567",
          base64[] = "0123456789abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ+/";
    char code[21]; /* Must be strictly superior than the highest code lentgh tested. */

    boringauth_hotp_init(&cfg);

    cfg.key = key;
    cfg.key_len = strlen(key);

    /* Base 10 */
    cfg.output_base = base10;
    cfg.output_base_len = strlen(base10);

    cfg.output_len = 5;
    uint32_t ret = boringauth_hotp_generate(&cfg, code);
    assert(ret == LIBREAUTH_OATH_CODE_TOO_SMALL);

    cfg.output_len = 6;
    ret = boringauth_hotp_generate(&cfg, code);
    assert(ret == LIBREAUTH_OATH_SUCCESS);
    assert(strlen(code) == 6);

    cfg.output_len = 9;
    ret = boringauth_hotp_generate(&cfg, code);
    assert(ret == LIBREAUTH_OATH_SUCCESS);
    assert(strlen(code) == 9);

    cfg.output_len = 10;
    ret = boringauth_hotp_generate(&cfg, code);
    assert(ret == LIBREAUTH_OATH_CODE_TOO_BIG);

    cfg.output_len = 0xffffff;
    ret = boringauth_hotp_generate(&cfg, code);
    assert(ret == LIBREAUTH_OATH_CODE_TOO_BIG);

    /* Base 32 */
    cfg.output_base = base32;
    cfg.output_base_len = strlen(base32);

    cfg.output_len = 3;
    ret = boringauth_hotp_generate(&cfg, code);
    assert(ret == LIBREAUTH_OATH_CODE_TOO_SMALL);

    cfg.output_len = 4;
    ret = boringauth_hotp_generate(&cfg, code);
    assert(ret == LIBREAUTH_OATH_SUCCESS);
    assert(strlen(code) == 4);

    cfg.output_len = 6;
    ret = boringauth_hotp_generate(&cfg, code);
    assert(ret == LIBREAUTH_OATH_SUCCESS);
    assert(strlen(code) == 6);

    cfg.output_len = 7;
    ret = boringauth_hotp_generate(&cfg, code);
    assert(ret == LIBREAUTH_OATH_CODE_TOO_BIG);

    cfg.output_len = 0xffffff;
    ret = boringauth_hotp_generate(&cfg, code);
    assert(ret == LIBREAUTH_OATH_CODE_TOO_BIG);

    /* Base 64 */
    cfg.output_base = base64;
    cfg.output_base_len = strlen(base64);

    cfg.output_len = 3;
    ret = boringauth_hotp_generate(&cfg, code);
    assert(ret == LIBREAUTH_OATH_CODE_TOO_SMALL);

    cfg.output_len = 4;
    ret = boringauth_hotp_generate(&cfg, code);
    assert(ret == LIBREAUTH_OATH_SUCCESS);
    assert(strlen(code) == 4);

    cfg.output_len = 5;
    ret = boringauth_hotp_generate(&cfg, code);
    assert(ret == LIBREAUTH_OATH_SUCCESS);
    assert(strlen(code) == 5);

    cfg.output_len = 6;
    ret = boringauth_hotp_generate(&cfg, code);
    assert(ret == LIBREAUTH_OATH_CODE_TOO_BIG);

    cfg.output_len = 0xffffff;
    ret = boringauth_hotp_generate(&cfg, code);
    assert(ret == LIBREAUTH_OATH_CODE_TOO_BIG);

    return 1;
}

uint32_t test_hotp(void) {
    uint32_t nb_tests = 0;

    nb_tests += test_basic_hotp();
    nb_tests += test_init_null_ptr();
    nb_tests += test_generate_null_ptr();
    nb_tests += test_invalid_base();
    nb_tests += test_invalid_code();

    return nb_tests;
}


tests/test_pass.c
/*
 * Copyright Rodolphe Breard (2016)
 * Author: Rodolphe Breard (<year>)
 *
 * This software is a computer program whose purpose is to [describe
 * functionalities and technical features of your software].
 *
 * This software is governed by the CeCILL  license under French law and
 * abiding by the rules of distribution of free software.  You can  use,
 * modify and/ or redistribute the software under the terms of the CeCILL
 * license as circulated by CEA, CNRS and INRIA at the following URL
 * "http://www.cecill.info".
 *
 * As a counterpart to the access to the source code and  rights to copy,
 * modify and redistribute granted by the license, users are provided only
 * with a limited warranty  and the software's author,  the holder of the
 * economic rights,  and the successive licensors  have only  limited
 * liability.
 *
 * In this respect, the user's attention is drawn to the risks associated
 * with loading,  using,  modifying and/or developing or reproducing the
 * software by the user in light of its specific status of free software,
 * that may mean  that it is complicated to manipulate,  and  that  also
 * therefore means  that it is reserved for developers  and  experienced
 * professionals having in-depth computer knowledge. Users are therefore
 * encouraged to load and test the software's suitability as regards their
 * requirements in conditions enabling the security of their systems and/or
 * data to be ensured and,  more generally, to use and operate it in the
 * same conditions as regards security.
 *
 * The fact that you are presently reading this means that you have had
 * knowledge of the CeCILL license and that you accept its terms.
 */


#include <assert.h>
#include <boringauth.h>
#include "boringauth_tests.h"


static uint32_t test_valid_pass(void) {
    test_name("pass: test_valid_pass");

    const char password[] = "correct horse battery staple",
          invalid_pass[] = "123456";
    uint8_t storage[LIBREAUTH_PASS_STORAGE_LEN];

    uint32_t ret = boringauth_pass_derive_password(password, storage, LIBREAUTH_PASS_STORAGE_LEN);
    assert(ret == LIBREAUTH_PASS_SUCCESS);
    assert(boringauth_pass_is_valid(password, storage));
    assert(!boringauth_pass_is_valid(invalid_pass, storage));

    return 1;
}

static uint32_t test_invalid_pass(void) {
    test_name("pass: test_invalid_pass");

    const char password[] = "invalid password",
          reference[] = "$pbkdf2-sha256$0$45217803$a607a72c2c92357a4568b998c5f708f801f0b1ffbaea205357e08e4d325830c9";
    assert(!boringauth_pass_is_valid(password, reference));

    return 1;
}

uint32_t test_pass(void) {
    int nb_tests = 0;

    nb_tests += test_valid_pass();
    nb_tests += test_invalid_pass();

    return nb_tests;
}


tests/test_totp.c
/*
 * Copyright Rodolphe Breard (2015)
 * Author: Rodolphe Breard (2015)
 *
 * This software is a computer program whose purpose is to [describe
 * functionalities and technical features of your software].
 *
 * This software is governed by the CeCILL  license under French law and
 * abiding by the rules of distribution of free software.  You can  use,
 * modify and/ or redistribute the software under the terms of the CeCILL
 * license as circulated by CEA, CNRS and INRIA at the following URL
 * "http://www.cecill.info".
 *
 * As a counterpart to the access to the source code and  rights to copy,
 * modify and redistribute granted by the license, users are provided only
 * with a limited warranty  and the software's author,  the holder of the
 * economic rights,  and the successive licensors  have only  limited
 * liability.
 *
 * In this respect, the user's attention is drawn to the risks associated
 * with loading,  using,  modifying and/or developing or reproducing the
 * software by the user in light of its specific status of free software,
 * that may mean  that it is complicated to manipulate,  and  that  also
 * therefore means  that it is reserved for developers  and  experienced
 * professionals having in-depth computer knowledge. Users are therefore
 * encouraged to load and test the software's suitability as regards their
 * requirements in conditions enabling the security of their systems and/or
 * data to be ensured and,  more generally, to use and operate it in the
 * same conditions as regards security.
 *
 * The fact that you are presently reading this means that you have had
 * knowledge of the CeCILL license and that you accept its terms.
 */


#include <strings.h>
#include <string.h>
#include <assert.h>
#include <boringauth.h>
#include "boringauth_tests.h"

#define DEFAULT_BUFF_LEN 6
#define BIGGER_BUFF_LEN  8


static uint32_t test_basic_totp(void) {
    test_name("totp: test_basic_totp");

    struct boringauth_totp_cfg cfg;
    const char key[] = "12345678901234567890";
    char code[DEFAULT_BUFF_LEN + 1] = {0};


    uint32_t ret = boringauth_totp_init(&cfg);
    assert(ret == LIBREAUTH_OATH_SUCCESS);
    assert(cfg.key == NULL);
    assert(cfg.key_len == 0);
    assert(cfg.timestamp != 0);
    assert(cfg.positive_tolerance == 0);
    assert(cfg.negative_tolerance == 0);
    assert(cfg.period == 30);
    assert(cfg.initial_time == 0);
    assert(cfg.output_len == DEFAULT_BUFF_LEN);
    assert(cfg.output_base == NULL);
    assert(cfg.output_base_len == 0);
    assert(cfg.hash_function == LIBREAUTH_OATH_SHA_1);

    cfg.key = key;
    cfg.key_len = sizeof(key);

    ret = boringauth_totp_generate(&cfg, code);
    assert(ret == LIBREAUTH_OATH_SUCCESS);
    assert(strlen(code) == DEFAULT_BUFF_LEN);
    assert(!boringauth_totp_is_valid(NULL, "755224"));
    assert(!boringauth_totp_is_valid(&cfg, "4755224"));
    assert(!boringauth_totp_is_valid(&cfg, "!@#$%^"));
    assert(!boringauth_totp_is_valid(&cfg, ""));
    assert(!boringauth_totp_is_valid(&cfg, NULL));

    return 1;
}

static uint32_t test_advanced_totp(void) {
    test_name("totp: test_advanced_totp");

    struct boringauth_totp_cfg cfg;
    const char key[] = "12345678901234567890123456789012";
    char code[BIGGER_BUFF_LEN + 1];

    uint32_t ret = boringauth_totp_init(&cfg);
    assert(ret == LIBREAUTH_OATH_SUCCESS);

    cfg.key = key;
    cfg.key_len = sizeof(key);
    cfg.timestamp = 1111111109;
    cfg.output_len = BIGGER_BUFF_LEN;
    cfg.hash_function = LIBREAUTH_OATH_SHA_256;

    ret = boringauth_totp_generate(&cfg, code);
    assert(ret == LIBREAUTH_OATH_SUCCESS);
    assert(strlen(code) == BIGGER_BUFF_LEN);
    assert(strncmp(code, "68084774", BIGGER_BUFF_LEN + 1) == 0);

    assert(boringauth_totp_is_valid(&cfg, "68084774"));
    assert(!boringauth_totp_is_valid(NULL, "68084774"));
    assert(!boringauth_totp_is_valid(&cfg, "68084775"));
    assert(!boringauth_totp_is_valid(&cfg, "46808477"));
    assert(!boringauth_totp_is_valid(&cfg, "!@#$%^&*"));
    assert(!boringauth_totp_is_valid(&cfg, ""));
    assert(!boringauth_totp_is_valid(&cfg, NULL));

    return 1;
}

static uint32_t test_tolerance(void) {
    test_name("totp: test_tolerance");

    struct boringauth_totp_cfg cfg;
    const char key[] = "12345678901234567890";

    uint32_t ret = boringauth_totp_init(&cfg);
    assert(ret == LIBREAUTH_OATH_SUCCESS);

    cfg.key = key;
    cfg.key_len = sizeof(key);
    cfg.timestamp = 1234567890;

    cfg.positive_tolerance = 0;
    cfg.negative_tolerance = 0;
    assert(!boringauth_totp_is_valid(&cfg, "590587"));
    cfg.positive_tolerance = 1;
    cfg.negative_tolerance = 1;
    assert(boringauth_totp_is_valid(&cfg, "590587"));
    cfg.positive_tolerance = 1;
    cfg.negative_tolerance = 1;
    assert(!boringauth_totp_is_valid(&cfg, "240500"));
    cfg.positive_tolerance = 2;
    cfg.negative_tolerance = 2;
    assert(boringauth_totp_is_valid(&cfg, "240500"));
    cfg.positive_tolerance = 0;
    cfg.negative_tolerance = 0;
    assert(!boringauth_totp_is_valid(&cfg, "980357"));
    cfg.positive_tolerance = 1;
    cfg.negative_tolerance = 1;
    assert(boringauth_totp_is_valid(&cfg, "980357"));
    cfg.positive_tolerance = 1;
    cfg.negative_tolerance = 1;
    assert(!boringauth_totp_is_valid(&cfg, "186057"));
    cfg.positive_tolerance = 2;
    cfg.negative_tolerance = 2;
    assert(boringauth_totp_is_valid(&cfg, "186057"));

    return 1;
}

static uint32_t test_init_null_ptr(void) {
    test_name("totp: test_init_null_ptr");

    uint32_t ret = boringauth_totp_init(NULL);
    assert(ret == LIBREAUTH_OATH_CFG_NULL_PTR);

    return 1;
}

static uint32_t test_generate_null_ptr(void) {
    test_name("totp: test_generate_null_ptr");

    struct boringauth_totp_cfg cfg;
    const char key[] = "12345678901234567890";
    char code[] = "qwerty";

    boringauth_totp_init(&cfg);

    uint32_t ret = boringauth_totp_generate(NULL, code);
    assert(ret == LIBREAUTH_OATH_CFG_NULL_PTR);
    assert(strcmp(code, "qwerty") == 0);

    ret = boringauth_totp_generate(&cfg, code);
    assert(ret == LIBREAUTH_OATH_KEY_NULL_PTR);

    cfg.key = key;

    ret = boringauth_totp_generate(&cfg, code);
    assert(ret == LIBREAUTH_OATH_INVALID_KEY_LEN);

    cfg.key_len = sizeof(key);

    ret = boringauth_totp_generate(&cfg, NULL);
    assert(ret == LIBREAUTH_OATH_CODE_NULL_PTR);

    ret = boringauth_totp_generate(&cfg, code);
    assert(ret == LIBREAUTH_OATH_SUCCESS);

    return 1;
}

static uint32_t test_invalid_base(void) {
    test_name("totp: test_invalid_base");

    struct boringauth_totp_cfg cfg;
    const char key[] = "12345678901234567890", base[] = "0123456789ABCDEF";
    char code[DEFAULT_BUFF_LEN + 1];

    boringauth_totp_init(&cfg);

    cfg.key = key;
    cfg.key_len = sizeof(key);
    cfg.output_base = base;

    uint32_t ret = boringauth_totp_generate(&cfg, code);
    assert(ret == LIBREAUTH_OATH_INVALID_BASE_LEN);
    cfg.output_base_len = 1;
    ret = boringauth_totp_generate(&cfg, code);
    assert(ret == LIBREAUTH_OATH_INVALID_BASE_LEN);

    cfg.output_base_len = sizeof(base);

    ret = boringauth_totp_generate(&cfg, code);
    assert(ret == LIBREAUTH_OATH_SUCCESS);

    return 1;
}

uint32_t test_totp(void) {
    uint32_t nb_tests = 0;

    nb_tests += test_basic_totp();
    nb_tests += test_advanced_totp();
    nb_tests += test_tolerance();
    nb_tests += test_init_null_ptr();
    nb_tests += test_generate_null_ptr();
    nb_tests += test_invalid_base();

    return nb_tests;
}


include/boringauth.h
/*
 *  This software is a computer program whose purpose is to compute validitiy of
 *  identification data.
 *
 *  Copyright (C) 2017 Th!nk Inc.
 *
 *  This program is free software; you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation, only version 2.0.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License along
 *  with this program; if not, write to the Free Software Foundation, Inc.,
 *  51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
 *
 */

/* Original LibreAuth License */

/*
 * Copyright Rodolphe Breard (2015)
 * Author: Rodolphe Breard (2015)
 *
 * This software is a computer program whose purpose is to [describe
 * functionalities and technical features of your software].
 *
 * This software is governed by the CeCILL  license under French law and
 * abiding by the rules of distribution of free software.  You can  use,
 * modify and/ or redistribute the software under the terms of the CeCILL
 * license as circulated by CEA, CNRS and INRIA at the following URL
 * "http://www.cecill.info".
 *
 * As a counterpart to the access to the source code and  rights to copy,
 * modify and redistribute granted by the license, users are provided only
 * with a limited warranty  and the software's author,  the holder of the
 * economic rights,  and the successive licensors  have only  limited
 * liability.
 *
 * In this respect, the user's attention is drawn to the risks associated
 * with loading,  using,  modifying and/or developing or reproducing the
 * software by the user in light of its specific status of free software,
 * that may mean  that it is complicated to manipulate,  and  that  also
 * therefore means  that it is reserved for developers  and  experienced
 * professionals having in-depth computer knowledge. Users are therefore
 * encouraged to load and test the software's suitability as regards their
 * requirements in conditions enabling the security of their systems and/or
 * data to be ensured and,  more generally, to use and operate it in the
 * same conditions as regards security.
 *
 * The fact that you are presently reading this means that you have had
 * knowledge of the CeCILL license and that you accept its terms.
 */

#pragma once

#include <stddef.h>
#include <stdint.h>


/*
 * PASS module
 */

#define LIBREAUTH_PASS_PASSWORD_MIN_LEN 4
#define LIBREAUTH_PASS_PASSWORD_MAX_LEN 128
#define LIBREAUTH_PASS_STORAGE_LEN      1024

typedef enum {
    LIBREAUTH_PASS_SUCCESS                  = 0,
    LIBREAUTH_PASS_PASSWORD_TOO_SHORT       = 1,
    LIBREAUTH_PASS_PASSWORD_TOO_LONG        = 2,
    LIBREAUTH_PASS_INVALID_PASSWORD_FORMAT  = 10,
    LIBREAUTH_PASS_NOT_ENOUGH_SPACE         = 20
} boringauth_pass_errno;

boringauth_pass_errno    boringauth_pass_derive_password(const void *password, void *storage, size_t storage_len);
int32_t                 boringauth_pass_is_valid(const void *password, const void *reference);


/*
 * OATH module
 */

typedef enum {
    LIBREAUTH_OATH_SHA_1   = 1,
    LIBREAUTH_OATH_SHA_256 = 2,
    LIBREAUTH_OATH_SHA_512 = 3
} boringauth_oath_hash_function;

typedef enum {
    LIBREAUTH_OATH_SUCCESS           = 0,

    LIBREAUTH_OATH_CFG_NULL_PTR      = 1,
    LIBREAUTH_OATH_CODE_NULL_PTR     = 2,
    LIBREAUTH_OATH_KEY_NULL_PTR      = 3,

    LIBREAUTH_OATH_INVALID_BASE_LEN  = 10,
    LIBREAUTH_OATH_INVALID_KEY_LEN   = 11,
    LIBREAUTH_OATH_CODE_TOO_SMALL    = 12,
    LIBREAUTH_OATH_CODE_TOO_BIG      = 13,

    LIBREAUTH_OATH_INVALID_KEY       = 20,
    LIBREAUTH_OATH_INVALID_PERIOD    = 21,

    LIBREAUTH_OATH_CODE_INVALID_UTF8 = 30
} boringauth_oath_errno;

/* HOTP */

struct boringauth_hotp_cfg {
    const void                   *key;
    size_t                        key_len;
    uint64_t                      counter;
    size_t                        output_len;
    const char                   *output_base;
    size_t                        output_base_len;
    boringauth_oath_hash_function  hash_function;
};

boringauth_oath_errno boringauth_hotp_init(struct boringauth_hotp_cfg *cfg);
boringauth_oath_errno boringauth_hotp_generate(const struct boringauth_hotp_cfg *cfg, char *code);
int32_t              boringauth_hotp_is_valid(const struct boringauth_hotp_cfg *cfg, const char *code);

/* TOTP */

struct boringauth_totp_cfg {
    const void                   *key;
    size_t                        key_len;
    int64_t                       timestamp;
    uint64_t                      positive_tolerance;
    uint64_t                      negative_tolerance;
    uint32_t                      period;
    uint64_t                      initial_time;
    size_t                        output_len;
    const void                   *output_base;
    size_t                        output_base_len;
    boringauth_oath_hash_function  hash_function;
};

boringauth_oath_errno boringauth_totp_init(struct boringauth_totp_cfg *cfg);
boringauth_oath_errno boringauth_totp_generate(const struct boringauth_totp_cfg *cfg, void *code);
int32_t              boringauth_totp_is_valid(const struct boringauth_totp_cfg *cfg, const void *code);
